### PR TITLE
376991: Create BIP rules

### PR DESCRIPTION
**Pull request by reece-bennett**
[AB#376991](https://dev.azure.com/defragovuk/54a51ea2-a7e4-4951-98ac-e063e18ccb63/_workitems/edit/376991)

### CODE REVIEW

File: src/STW.ProcessingApi.Function/Triggers/ProcessingApiTrigger.cs [hunk 1]
```diff
@@ -19,12 +19,31 @@ public ProcessingApiTrigger(ILogger<ProcessingApiTrigger> logger, IValidationSer
...
     {
         _logger.LogInformation($"{nameof(ProcessingApiTrigger)} function was invoked.");
 
-        var spsCertificate = JsonSerializer.Deserialize<SpsCertificate>(message.Body);
+        var messageBody = message.Body.ToString();
 
-        await _validationService.InvokeRulesAsync(spsCertificate!);
+        try
+        {
+            var spsCertificate = JsonConvert.DeserializeObject<SpsCertificate>(messageBody);
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Triggers/ProcessingApiTrigger.cs]: Why are you using Newtonsoft here over `System.Text.Json`?

File: src/STW.ProcessingApi.Function/Triggers/ProcessingApiTrigger.cs [hunk 2]
```diff
@@ -19,12 +19,31 @@ public ProcessingApiTrigger(ILogger<ProcessingApiTrigger> logger, IValidationSer
...
     [Function(nameof(ProcessingApiTrigger))]
-    public async Task Run([ServiceBusTrigger("%ServiceBusQueueName%", Connection = "ServiceBusConnectionString")] ServiceBusReceivedMessage message)
+    public async Task Run(
+        [ServiceBusTrigger("%ServiceBusQueueName%", Connection = "ServiceBusConnectionString")]
+        ServiceBusReceivedMessage message)
     {
         _logger.LogInformation($"{nameof(ProcessingApiTrigger)} function was invoked.");
 
-        var spsCertificate = JsonSerializer.Deserialize<SpsCertificate>(message.Body);
+        var messageBody = message.Body.ToString();
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Triggers/ProcessingApiTrigger.cs]: Do we 100% need the .ToString()? It was working without it beforehand.

File: src/STW.ProcessingApi.Function/Triggers/ProcessingApiTrigger.cs [hunk 3]
```diff
@@ -19,12 +19,31 @@ public ProcessingApiTrigger(ILogger<ProcessingApiTrigger> logger, IValidationSer
...
+        try
+        {
+            var spsCertificate = JsonConvert.DeserializeObject<SpsCertificate>(messageBody);
+
+            var errors = await _validationService.InvokeRulesAsync(spsCertificate!);
+
+            if (errors.Count == 0)
+            {
+                _logger.LogInformation("Validation passed");
+            }
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Triggers/ProcessingApiTrigger.cs]: The validation passed/failed logic would be better placed in the validation service rather than here.

I don't think we will need to catch the JSON exception and should let it bubble up as it will be visible in the Azure portal.

File: src/STW.ProcessingApi.Function/Services/ValidationService.cs [hunk 4]
```diff
@@ -18,7 +18,7 @@ public ValidationService(IEnumerable<IRule> rules, IEnumerable<IAsyncRule> async
         _logger = logger;
     }
 
-    public async Task InvokeRulesAsync(SpsCertificate spsCertificate)
+    public async Task<List<ValidationError>> InvokeRulesAsync(SpsCertificate spsCertificate)
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Services/ValidationService.cs]: See other comment on trigger and change this back.

File: src/STW.ProcessingApi.Function/Services/BcpService.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+    public BcpService(HttpClient httpClient)
+    {
+        _httpClient = httpClient;
+    }
+
+    public async Task<List<Bcp>> GetBcpsWithCodeAndType(string code, string chedType)
+    {
+        var response =
+            await _httpClient.GetFromJsonAsync<BcpSearchResponse>($"/bcps/search?code={code}&type={chedType}");
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Services/BcpService.cs]: You should use `GetAsync` instead and add appropriate error handling

File: src/STW.ProcessingApi.Function/STW.ProcessingApi.Function.csproj [hunk 6]
```diff
@@ -6,11 +6,12 @@
...
-        <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.19.0"/>
+        <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.22.0" />
+        <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Http" Version="3.2.0" />
         <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.ServiceBus" Version="5.18.0" />
-        <PackageReference Include="Microsoft.Azure.Functions.Worker.Sdk" Version="1.15.1"/>
-        <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Http" Version="3.1.0"/>
-        <PackageReference Include="Microsoft.Extensions.Diagnostics.HealthChecks" Version="8.0.4" />
+        <PackageReference Include="Microsoft.Azure.Functions.Worker.Sdk" Version="1.17.2" />
+        <PackageReference Include="Microsoft.Extensions.Diagnostics.HealthChecks" Version="8.0.6" />
+        <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/STW.ProcessingApi.Function.csproj]: No need for Newtonsoft

File: src/STW.ProcessingApi.Function/Rules/Common/BcpValidRule.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+{
+    private readonly IBcpService _bcpService;
+
+    public BcpValidRule(IBcpService bcpService)
+    {
+        _bcpService = bcpService;
+    }
+
+    public bool ShouldInvoke(SpsCertificate spsCertificate)
+    {
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Rules/Common/BcpValidRule.cs]: This can be changed to `public bool ShouldInvoke(SpsCertificate spsCertificate) => true;` if you like.

File: src/STW.ProcessingApi.Function/Rules/Common/BcpValidRule.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        return true;
+    }
+
+    public async Task InvokeAsync(SpsCertificate spsCertificate, IList<ValidationError> errors)
+    {
+        var chedType = SpsCertificateHelper.GetChedType(spsCertificate.SpsExchangedDocument.IncludedSpsNote) ??
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Rules/Common/BcpValidRule.cs]: Don't use `string.Empty` everywhere. We know that CHED TYPE will be present so do `var chedType = SpsCertificateHelper.GetChedType(spsCertificate.SpsExchangedDocument.IncludedSpsNote);` and use `chedType!` in the params.

File: src/STW.ProcessingApi.Function/Extensions/ServiceCollectionExtensions.cs [hunk 9]
```diff
@@ -25,14 +28,26 @@ public static void RegisterRules(this IServiceCollection serviceCollection)
...
         }
     }
+
+    public static void RegisterHttpClients(this IServiceCollection serviceCollection)
+    {
+        serviceCollection.AddHttpClient<IBcpService, BcpService>(
+            (serviceProvider, httpClient) =>
+            {
+                var options = serviceProvider.GetRequiredService<IOptions<ApiConfigOptions>>().Value;
+                httpClient.BaseAddress = new Uri(options.BcpServiceBaseUrl);
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Extensions/ServiceCollectionExtensions.cs]: Set the timeout on the client here also

File: src/STW.ProcessingApi.Function.UnitTests/Rules/Common/BcpValidRuleTest.cs [hunk 10]
```diff
@@ -0,0 +1,106 @@
+namespace STW.ProcessingApi.Function.UnitTests.Rules.Common;
+
+using FluentAssertions;
+using Function.Rules.Common;
+using Function.Services;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Models;
+using Moq;
+using Newtonsoft.Json;
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function.UnitTests/Rules/Common/BcpValidRuleTest.cs]: Remove Newtonsoft in favour of built-in package.

File: src/STW.ProcessingApi.Function.UnitTests/Services/BcpServiceTest.cs [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+            .ReturnsHttpResponseAsync(
+                new BcpSearchResponse
+                {
+                    Bcps = [bcp],
+                },
+                HttpStatusCode.OK);
+
+        var httpClient = new HttpClient(httpMessageHandler.Object)
+        {
+            BaseAddress = new Uri("https://bcp.service"),
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function.UnitTests/Services/BcpServiceTest.cs]: Introduce a [TestInitialize] annotation and method to this class and instantiate your system under test, HTTP client and mocks there rather than duplicate them.

File: src/STW.ProcessingApi.Function/Models/ValidationError.cs [hunk 12]
```diff
@@ -2,16 +2,51 @@ namespace STW.ProcessingApi.Function.Models;
...
+    public int? ErrorId { get; }
+
+    public int? ErrorTradeLineItem { get; }
+
+    public override string ToString()
+    {
+        return $"{ErrorMessage}";
+    }
+
+    public override bool Equals(object? obj)
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Models/ValidationError.cs]: Unless you use any of this anywhere, I would recommend removing it.

File: src/STW.ProcessingApi.Function/Services/Interfaces/IValidationService.cs [hunk 13]
```diff
@@ -4,5 +4,5 @@ namespace STW.ProcessingApi.Function.Services.Interfaces;
 
 public interface IValidationService
 {
-    Task InvokeRulesAsync(SpsCertificate spsCertificate);
+    Task<List<ValidationError>> InvokeRulesAsync(SpsCertificate spsCertificate);
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Services/Interfaces/IValidationService.cs]: No need for this to return the list of errors.

File: src/STW.ProcessingApi.Function/Triggers/ProcessingApiTrigger.cs (hunk 3 above)
Comment (reece-bennett) [reply src/STW.ProcessingApi.Function/Triggers/ProcessingApiTrigger.cs]: Feels like it makes more sense to have the handling of the errors here for when they get sent to another service so we can expose them to STW? I'll change back for now

File: src/STW.ProcessingApi.Function/Services/BcpService.cs (hunk 5 above)
Comment (reece-bennett) [reply src/STW.ProcessingApi.Function/Services/BcpService.cs]: This works fine - it throws an exception if it's non-2xx status and if there is a network error (e.g. service down) which is then handled in the rule

File: src/STW.ProcessingApi.Function.UnitTests/Services/BcpServiceTest.cs (hunk 11 above)
Comment (reece-bennett) [reply src/STW.ProcessingApi.Function.UnitTests/Services/BcpServiceTest.cs]: They change in each test

File: src/STW.ProcessingApi.Function/Rules/Common/BcpValidRule.cs (hunk 7 above)
Comment (sean-treanor) [reply src/STW.ProcessingApi.Function/Rules/Common/BcpValidRule.cs]: It might be better for clarity and a unit testing perspective which certificate types the rule applies to.

File: src/STW.ProcessingApi.Function/Services/BcpService.cs (hunk 5 above)
Comment (sean-treanor) [reply src/STW.ProcessingApi.Function/Services/BcpService.cs]: Not sure if exception handling in the rule itself is the best approach for this. Might be better looking into the Result wrapper approach that I have implemented. We likely want consistency between the rules?

File: src/STW.ProcessingApi.Function.UnitTests/Services/BcpServiceTest.cs (hunk 11 above)
Comment (sean-treanor) [reply src/STW.ProcessingApi.Function.UnitTests/Services/BcpServiceTest.cs]: The mock instantiation and HTTP Client don't change? Please take a look at this test class:
https://github.com/DEFRA/stw-processing-api/blob/a0251a95fa2eb1d7c4b0684c75e201d1bfbb6261/src/STW.ProcessingApi.Function.UnitTests/Services/ApprovedEstablishmentServiceTests.cs

File: src/STW.ProcessingApi.Function.UnitTests/TestExtensions/MoqExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+public static class MoqExtensions
+{
+    public static ISetup<HttpMessageHandler, Task<HttpResponseMessage>> SetupSendAsync(
+        this Mock<HttpMessageHandler> handler,
+        HttpMethod requestMethod,
+        string requestUrl)
+    {
+        return handler.Protected().Setup<Task<HttpResponseMessage>>(
+            "SendAsync",
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function.UnitTests/TestExtensions/MoqExtensions.cs]: There are Moq extensions already that I added that let you do this. Might be best removing these in favour of those.

File: src/STW.ProcessingApi.Function/Constants/ChedType.cs [hunk 15]
```diff
@@ -2,6 +2,16 @@ namespace STW.ProcessingApi.Function.Constants;
 
 public static class ChedType
 {
-    public const string Chedpp = "CHED-PP";
-    public const string Chedp = "CHED-P";
+    public const string Cheda = "CHEDA";
+    public const string Chedp = "CHEDP";
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Constants/ChedType.cs:6]: Why did you change CHED-PP and CHED-P to CHEDPP and CHEDP? Both the schema we were working from and the notification schema have these original values.

File: src/STW.ProcessingApi.Function/Constants/ChedType.cs [hunk 16]
```diff
@@ -2,6 +2,16 @@ namespace STW.ProcessingApi.Function.Constants;
...
+    public const string Chedd = "CHEDD";
+    public const string Chedpp = "CHEDPP";
+
+    public static readonly List<string> Values =
+    [
+        Cheda,
+        Chedp,
+        Chedd,
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Constants/ChedType.cs:14]: Do we need CHEDA and CHEDD? Given you have added these to the values list and reference this in your common rules, these cert types can be used to trigger rules?

Comment (sean-treanor) [reply src/STW.ProcessingApi.Function/Constants/ChedType.cs:14]: I was under the impression these rules only apply to CHEDP and CHEDPP

File: src/STW.ProcessingApi.Function/Constants/RuleErrorId.cs [hunk 17]
```diff
@@ -48,4 +48,7 @@ public static class RuleErrorId
     public const int InvalidRegionOfOrigin = 86;
     public const int MoreThanOneCountryOfOrigin = 87;
     public const int MoreThanOneRegionOfOriginInConsignment = 88;
+    public const int InvalidBcpCode = 89;
+    public const int BcpSuspended = 90;
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Constants/RuleErrorId.cs]: These codes overlap with ones in main

File: src/STW.ProcessingApi.Function/Rules/Common/BcpValidRule.cs [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            if (result.Value?.Count == 0)
+            {
+                errors.Add(
+                    new ValidationError(
+                        string.Format(RuleErrorMessage.InvalidBcpCode, bcpCode, chedType),
+                        RuleErrorId.InvalidBcpCode));
+            }
+            else if (result.Value![0].Suspended)
+            {
```
Comment (sean-treanor) [inline src/STW.ProcessingApi.Function/Rules/Common/BcpValidRule.cs]:
```csharp
if (!result.isSuccess)
{
  errors.Add(new ValidationError(string.Format(RuleErrorMessage.BcpServiceError, result.Error.Message), RuleErrorId.BcpServiceError));

  return;
}

if (result.Value!.Count == 0)
{
  errors.Add(new ValidationError(string.Format(RuleErrorMessage.InvalidBcpCode, bcpCode, chedType), RuleErrorId.InvalidBcpCode));

  return;
}

if (result.Value![0].Suspended)
{
  errors.Add(new ValidationError(string.Format(RuleErrorMessage.BcpSuspended, bcpCode, chedType), RuleErrorId.BcpSuspended));
}
```

You could remove nested complexity from this logic by taking a different approach.

File: src/STW.ProcessingApi.Function/Constants/ChedType.cs (hunk 15 above)
Comment (reece-bennett) [reply src/STW.ProcessingApi.Function/Constants/ChedType.cs:6]: It's what was agreed for the payload format, see the wiki https://dev.azure.com/defragovuk/DEFRA-STW-Integration/_wiki/wikis/STW-Integration.wiki/23404/Common-Payload-Description, and it's the value used in our version of the schema https://github.com/DEFRA/stw-mapping-library/blob/main/trade-schema/src/main/resources/json/enotification/definitions/IncludedSpsNote.json#L90

File: src/STW.ProcessingApi.Function/Constants/ChedType.cs (hunk 16 above)
Comment (reece-bennett) [reply src/STW.ProcessingApi.Function/Constants/ChedType.cs:14]: The orchestration layer needs to work for all certificate types

**Review by sean-treanor — APPROVED** [review]

**Review by kamilmkainos — APPROVED** [review]
